### ISSUE TITLE
Constructors not filtered out of documentation by access level

### ISSUE THREAD

**Issue by bjnsn** [jleyba/js-dossier#26]
As I understand it, the following should not be outputted to the documentation due to @private access. 

    /**
    * My hidden Foo
    * @private
    * @constructor
    */
     my.Foo = function() { };

It would be helpful to have constructors filtered out by access. 

In my case something like the [http://usejsdoc.org/about-commandline.html](JSDOC -a/--access command-line option) would be useful since I'm developing a component. I would tend to use @protected heavily so that only the interface is exposed in the docs.

**Comment by jleyba** [comment]
Constructors are tricky. Marking them as `@private` dictates that the constructor may only be invoked from within the same file, not that the class itself is restricted to that file. This lets you restrict how a class may be created (e.g. factory functions).  Here's an example:

http://closure-compiler-debugger.appspot.com/#input0%3Dgoog.provide('foo')%253B%250A%250A%250A%252F**%250A%2520*%2520%2540private%250A%2520*%2520%2540constructor%250A%2520*%252F%250Afoo.Bar%2520%253D%2520function()%2520%257B%257D%253B%250A%250A%250A%252F**%2520%2540return%2520%257B!foo.Bar%257D%2520.%2520*%252F%250Afoo.createBar%2520%253D%2520function()%2520%257B%250A%2520%2520return%2520new%2520foo.Bar()%253B%250A%257D%253B%26input1%3Dgoog.require('foo')%253B%250A%250A%252F**%250A%2520*%2520%2540param%2520%257B!foo.Bar%257D%2520value%2520.%250A%2520*%252F%250Afunction%2520check(value)%2520%257B%250A%257D%250A%250A%252F%252F%2520This%2520is%2520OK.%250Avar%2520bar%2520%253D%2520foo.createBar()%253B%250Acheck(bar)%253B%250A%250A%250A%252F%252F%2520This%2520is%2520not%2520OK.%250Anew%2520foo.Bar()%253B%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CHECK_VISIBILITY%3D1%26CLOSURE_PASS%3D1%26MISSING_PROPERTIES%3D1


Maybe a good compromise would be: if all of the references to a `@private` constructor are contained within its containing file, consider the class private and omit from generated documentation. Otherwise, consider the constructor itself private and include the class in generated documentation.

**Comment by jleyba** [comment]
Raising the visibility of the remaining tasks listed commit d0fa6af:

> There should be a config option to control what level of visibility to document (public, protected, or private). Furthermore, logic needs to be added for the inherited case, where a visibility has been set at the file level.

**Comment by myphysicslab** [comment]
The root of the problem is overloading the visibility of the class and the constructor into one place: the constructor's visibility.  There are various ways to deal with this.

1.  The constructor visibility affects only the visibility of the constructor.  Classes are always visible unless they are listed in the "excludes" portion of the `dossier_config.json` file.

2.  The constructor visibility specifies the visibility of the class.  This prohibits the important usages like Singleton pattern or Utility class as discussed in issue #36.

3.  Add some marker to the class about it's visibility.  For example, define a property on the class called `__dossier_private__` to make the whole class private.

Note that the notion of an overall "level of visibility to document" is orthogonal to (separate from) this.

In particular, looking at whether a `@private` constructor is used within its file will not work with the Utility class discussed in issue #36, because those are static classes that are never instantiated.

I vote for option 1.

**Comment by jleyba** [comment]
Yes, I agree that option 1 is probably the best course for constructors. This should also apply to interfaces so any type references to the interface are properly resolved for links.

I think we should continue to exclude `@private` properties and functions (for now).

**Comment by myphysicslab** [comment]
Excluding `@private` properties and functions works well for me now.

It would be nice to either have the option to exclude `@protected` properties and functions, or failing that perhaps some indication that they are `@protected`.

**Comment by jleyba** [comment]
Definitely. An older version of Dossier did this. Here's an [example](https://selenium.googlecode.com/git-history/selenium-2.40.0/docs/api/javascript/class_webdriver_WebDriver.html).

I'd like to do something like this again, just haven't figured out the best way yet.

**Comment by myphysicslab** [comment]
I'm still seeing that constructors marked `@private` **are** appearing in the documentation -- it would be more consistent to hide the constructor because that's what is done now for `@private` properties and functions.  

But it's good that overall the class is back to being documented.  (And of course the class description from the constructor should always be there even when the constructor is hidden).
